Triangle-cross arrowhead

d2fong opened the pull request:
## High Level Overview

- implements a new arrow head shape called 'triangle-cross'
- collision detection using inline vicinity
- scales arrow size based on edge width and size


## Miscellaneous Changes

- removes explicit arrowShape check for 'half-triangle-overshot' in the drawArrowHead function
- reuses arrowWidth and arrowShape variables that were already computed at the beginning of the drawArrowHead function
- shapeImpl.draw takes edgeWidth as a parameter for arrow scaling

See https://github.com/cytoscape/cytoscape.js/issues/1656

d2fong commented on src/extensions/renderer/base/arrow-shapes.js:
```diff
@@ -191,6 +191,56 @@ BRp.registerArrowShapes = function(){
+    },
+    scaleCrossLineXCoord: function( size, edgeWidth ){
+      return 0.42 + ( edgeWidth * 0.012 ) +  ( math.log2( size - 28.95 ) * 0.001 );
+    },
```
These two functions could be changed to just one, with the costant as a parameter, but I thought it would be more clear that both the x and y coordinates of the cross can be scaled independently.

maxkfranz replied:
Maybe `scaleCrossLineYCoord` and X can just reference the same impl of a function defined in a block above?

d2fong commented on src/extensions/renderer/base/arrow-shapes.js:
```diff
@@ -191,6 +191,56 @@ BRp.registerArrowShapes = function(){
+      return inside;
+    },
+
+    draw: function( context, size, angle, translation, edgeWidth ){
```
I found that I could make the scaling functions more accurate if I used the edgeWidth parameter to calculate the scaling.

d2fong commented on src/extensions/renderer/canvas/drawing-edges.js:
```diff
@@ -181,45 +181,42 @@ CRp.drawArrowheads = function( context, edge, drawOverlayInstead ){
 CRp.drawArrowhead = function( context, edge, prefix, x, y, angle ){
   if( isNaN( x ) || x == null || isNaN( y ) || y == null || isNaN( angle ) || angle == null ){ return; }
 
-  var self = this;
```
I don't think self is needed in this case?

maxkfranz replied:
OK.  It's probably just used in most functions in case a ref to `this` is needed in an inner function.  It may not be needed here, but a side benefit is smaller minified files.  `this` is always 4 chars minified because it's a keyword; `self` becomes something 1 char like `x`.

d2fong commented on src/extensions/renderer/canvas/drawing-edges.js:
```diff
@@ -181,45 +181,42 @@ CRp.drawArrowheads = function( context, edge, drawOverlayInstead ){
+  var arrowWidth = edge.pstyle( 'width' ).pfValue;
   var opacity = edge.pstyle( 'opacity' ).value;
 
-  if( arrowShape === 'half-triangle-overshot' ){
```
Removes old code like you said.

d2fong commented on src/extensions/renderer/canvas/drawing-edges.js:
```diff
@@ -181,45 +181,42 @@ CRp.drawArrowheads = function( context, edge, drawOverlayInstead ){
-      arrowClearFill, edge.pstyle( 'width' ).pfValue, edge.pstyle( prefix + '-arrow-shape' ).value,
-      x, y, angle
+    this.drawArrowShape( edge, prefix, context,
+      arrowClearFill, arrowWidth, arrowShape, x, y, angle
```
Uses arrowWidth, arrowShape variables instead of getting the value again.

d2fong commented on src/extensions/renderer/canvas/drawing-edges.js:
```diff
@@ -253,7 +250,7 @@ CRp.drawArrowShape = function( edge, arrowType, context, fill, edgeWidth, shape,
 
   if( !pathCacheHit ){
-    shapeImpl.draw( context, size, angle, translation );
+    shapeImpl.draw( context, size, angle, translation, edgeWidth );
```
This is where I send the edgeWidth to the 'triangle-cross' draw function.  Again, thoughts on this?

maxkfranz commented on src/extensions/renderer/canvas/drawing-edges.js:
```diff
@@ -181,45 +181,42 @@ CRp.drawArrowheads = function( context, edge, drawOverlayInstead ){
 
   var arrowClearFill = edge.pstyle( prefix + '-arrow-fill' ).value === 'hollow' ? 'both' : 'filled';
   var arrowFill = edge.pstyle( prefix + '-arrow-fill' ).value;
+  var arrowWidth = edge.pstyle( 'width' ).pfValue;
```
Should not be called `arrowWidth`.  That's actually a different value, semantically.  Better to call it `edgeWidth` or it will be confusing.

d2fong commented on src/extensions/renderer/base/arrow-shapes.js:
```diff
@@ -191,6 +191,60 @@ BRp.registerArrowShapes = function(){
+
+    scaleCoord: function ( constant, size, edgeWidth ){
+     return constant + ( edgeWidth * 0.012 ) + ( math.log2( size - 28.95 ) * 0.001 );
+    },
```
base function to scale a coordinate

d2fong commented on src/extensions/renderer/canvas/drawing-edges.js:
```diff
@@ -183,20 +183,20 @@ CRp.drawArrowhead = function( context, edge, prefix, x, y, angle ){
 
   var arrowClearFill = edge.pstyle( prefix + '-arrow-fill' ).value === 'hollow' ? 'both' : 'filled';
   var arrowFill = edge.pstyle( prefix + '-arrow-fill' ).value;
+  var edgeWidth = edge.pstyle( 'width' ).pfValue;
```
renamed variable to edgewidth

d2fong commented:
@maxkfranz  I have made the changes that you have suggested. 

List of Changes:
- Reverted the self -> this change
- Renamed the incorrect arrowWidth name to edgeWidth 
- Renamed variables with prefix 'tee' to the prefix 'crossLine' in the triangle-cross implementation

This new arrow head scales both with edge width and arrow scale.  When both values are large, it makes it look a bit wonky, but I believe it is working.  EdgeWidth scales the edge width, arrow scale scales the arrow head.
